source/common/tcp/async_tcp_client_impl.cc

```diff
@@ -24,6 +24,14 @@ AsyncTcpClientImpl::AsyncTcpClientImpl(Event::Dispatcher& dispatcher,
       connect_timer_(dispatcher.createTimer([this]() { onConnectTimeout(); })),
       enable_half_close_(enable_half_close) {}
 
+AsyncTcpClientImpl::~AsyncTcpClientImpl() {
+  if (connection_) {
+    connection_->removeConnectionCallbacks(*this);
+  }
+
+  close(Network::ConnectionCloseType::NoFlush);
```

Comment:
Could you elaborate more for the reason that the `close()` is added here? I don't see from your initial fix, and  I feel a user is possible to close the connection with other type.

Reply:
The reason is because if we don't close it, tests fail because of the ``ASSERT`` in:
https://github.com/envoyproxy/envoy/blob/b1e3351d7f3147afab9d29596d324d93e2198ba3/source/common/network/connection_impl.cc#L116-L125

The comment in that code section explains the expectation out of the owner of the ``ConnectionImpl``:
``In general we assume that owning code has called close() previously to the destructor being run.``

Reply:
I feel this is to mitigate the wrong usage of this client in the test filter to simulate the callback issue, and ideally this should not happen. It needs to be fixed in the usage part. FWIW, the original version of fix LGTM, and we can change how we test it or remove the newlly added tests.

Reply:
The changes differ in the ``close`` call, is there a reason why we shouldn't include it, regardless of the testing?
Removing the newly added test meaning no integration test coverage for this case.

Reply:
I think the reason you see the non-close error is because we force release the async_client before calling any close as the wrong usage, and that is why we have the assertion. If we really want to keep the close here to cover this should-not-happen case, adding a close boolean for close may be less risky. I am more worried that if a caller calls `close(FlushWriteAndDelay)` and then following with a `close(NoFlush)` when the connection is still in the defer deletion list.